File: small_problems/easy8/all_substrings.rb
# Write a method that returns a list of all substrings of a string. The returned list should be ordered by where in the string the substring begins. This means that all substrings that start at position 0 should come first, then all substrings that start at position 1, and so on. Since multiple substrings will occur at each position, the substrings at a given position should be returned in order from shortest to longest.

# You may (and should) use the leading_substrings method you wrote in the previous exercise:

def substrings(string)
  array = Array.new

  string.chars.each_with_index do |_, index|
    1.upto(string.size - index) do |count|
      array << string[index, count]
    end
  end

  array
end

# alternate

def leading_substrings(string)
  string.chars.map.with_index do |_, index|
    string[0..index]
  end
end

def substrings(string)
  array = []

  (string.size).times do |index|
    array.concat(leading_substrings(string[index..-1]))
  end

  array
end

p substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

File: javascript_small_problems/list_processing/all_substrings.js
function substrings(string) {
  let result = [];
  for (let index1 = 0; index1 < string.length; index1 += 1) {
    for (let index2 = index1; index2 < string.length; index2 += 1) {
      result.push(string.slice(index1, index2 + 1));
    }
  }

  return result;
}

// Further exploration
function substrings(string) {
  return [...string].flatMap((_, index1) => {
    let substring = string.slice(index1);
    return [...substring].map((_, index2) => substring.slice(0, index2 + 1))
  })
}

console.log(substrings('abcde'));

// returns
// [ "a", "ab", "abc", "abcd", "abcde",
//   "b", "bc", "bcd", "bcde",
//   "c", "cd", "cde",
//   "d", "de",
//   "e" ]